src/main/java/de/tum/in/www1/artemis/security/lti/Lti13TokenRetriever.java

```diff
@@ -0,0 +1,127 @@
+package de.tum.in.www1.artemis.security.lti;
+
+import java.net.URI;
+import java.security.KeyPair;
+import java.time.Instant;
+import java.util.*;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.*;
+import org.springframework.security.oauth2.client.registration.ClientRegistration;
+import org.springframework.stereotype.Component;
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.client.HttpClientErrorException;
+import org.springframework.web.client.RestTemplate;
+
+import com.google.gson.JsonParser;
+import com.nimbusds.jose.JOSEException;
+import com.nimbusds.jose.JOSEObjectType;
+import com.nimbusds.jose.JWSAlgorithm;
+import com.nimbusds.jose.JWSHeader;
+import com.nimbusds.jose.crypto.RSASSASigner;
+import com.nimbusds.jose.jwk.JWK;
+import com.nimbusds.jwt.JWTClaimsSet;
+import com.nimbusds.jwt.SignedJWT;
+
+import de.tum.in.www1.artemis.security.OAuth2JWKSService;
+
+/**
+ * This class is responsible to retrieve access tokens from an LTI 1.3 platform of a specific ClientRegistration.
+ *
+ */
+@Component
+public class Lti13TokenRetriever {
+
+    private final OAuth2JWKSService oAuth2JWKSService;
+
+    private final RestTemplate restTemplate;
+
+    private final Logger log = LoggerFactory.getLogger(Lti13TokenRetriever.class);
+
+    private static final int JWT_LIFETIME = 60;
+
+    public Lti13TokenRetriever(OAuth2JWKSService keyPairService) {
+        this.oAuth2JWKSService = keyPairService;
+        this.restTemplate = new RestTemplate();
+    }
+
+    /**
+     * Returns an access token for a specific client to be used to authenticate further communication with the related
+     * LTI 1.3 platform of that client. The queried access token (if there is any) will be valid for the requested scopes.
+     *
+     * @param clientRegistration to query a token for
+     * @param scopes to ask access for
+     * @return the access token to be used to authenticate requests to the client's LTI 1.3 platform.
+     */
+    public String getToken(ClientRegistration clientRegistration, String... scopes) {
+        log.info("Trying to retrieve access token for client");
+        try {
```

Comment:
scope try the try block to the critical parts

Reply:
Done in [a141aff](https://github.com/ls1intum/Artemis/pull/5693/commits/a141afff2df027fa6a99788faeb523e55fdb7eab) as well